src/workerd/io/worker-entrypoint.c++

```diff
@@ -213,6 +213,63 @@ void WorkerEntrypoint::init(
       .attach(kj::mv(actor));
 }
 
+/**
+ * Setup Python with a more generous limit enforcer, currently enterStartupJs.
+ *
+ * The Pythons setup is async so we hold the limitScope until the setup work is done. Whatever else
+ * happens between now and when we finish initPython will be done with the higher limit. User code
+ * cannot be entered with the higher limit because the first thing the main entrypoints do is wait
+ * on this initialization work being completed.
+ *
+ * Ideally we should be able to wrap a JS task so that only it is executed with a different limiter,
+ * which is switched on when entering the task and back off when exiting. But I don't know how to do
+ * that so maybe this works okay for now.
+ */
+void
+initPythonIfNeeded(kj::Maybe<kj::StringPtr>& entrypointName, kj::Own<workerd::IoContext_IncomingRequest>& incomingRequest) {
+  auto& context = incomingRequest->getContext();
+  auto& script = context.getWorker().getScript();
+  if (!script.isPython) {
+    return;
+  }
+  context.addWaitUntil(context.run(
+      [entrypointName=entrypointName, &context, &metrics = incomingRequest->getMetrics()]
+      (Worker::Lock& lock) mutable -> kj::Promise<void> {
+    kj::Maybe<kj::Exception> maybeLimitError;
+
+    // Raise limit
+    auto& isolate = Worker::Isolate::from(lock);
+    kj::Own<void> limitScope = isolate.getLimitEnforcer().enterStartupJs(lock, maybeLimitError);
+
+    // call initPython
+    auto handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, context.getActor()), "Failed to get worker handler.");
+    jsg::Promise<void> res = nullptr;
+    KJ_IF_SOME(initPython, handler->initPython) {
+      res = initPython(lock, handler->getCtx());
+    } else {
+      KJ_FAIL_REQUIRE("Internal error, expected to see initPython on handler.");
+    }
+
+    // Goal: attach a callback to the promise so that limitScope isn't dropped until the callback
+    // returns. We need the owner of limitScope to be an IoOwn since we seem to need the lock to
+    // drop limitScope.
+    //
+    // Unfortunately the type of limitScope is kj::Own<void> and IoContext::addObject does not work
+    // on a kj::Own<void>. Work around this by making a LimitScopeHolder class that holds the
+    // limitScope and then attach the limitScopeHolder to the IoContext.
+    class LimitScopeHolder {
+    public:
+      kj::Own<void> limitScope;
+      LimitScopeHolder(kj::Own<void> limitScope) : limitScope(kj::mv(limitScope)) {}
+    };
+    auto limitScopeHolder = IoContext::current().addObject(kj::heap<LimitScopeHolder>(kj::mv(limitScope)));
```

Comment:
Regardless of whether overlapping CPU limits work, extending the CPU limit's lifetime beyond the scope of a (synchronous) function call is also going to create trouble. We can't be returning to the event loop with a CPU limit still in force.